src/audio/volume/volume.c

```diff
@@ -55,6 +56,139 @@ DECLARE_SOF_UUID("volume", volume_uuid, 0xb77e677e, 0x5ff4, 0x4188,
 DECLARE_SOF_UUID("volume-task", volume_task_uuid, 0x34dc0385, 0xfc2f, 0x4f7f,
 		 0x82, 0xd2, 0x6c, 0xee, 0x44, 0x45, 0x33, 0xe0);
 
+#if CONFIG_FORMAT_S16LE
+/**
+ * \brief Used to find nearest zero crossing frame for 16 bit format.
+ * \param[in,out] source Source buffer.
+ * \param[in] frames Number of frames.
+ */
+static uint32_t vol_zc_get_s16(const struct audio_stream *source,
+			       uint32_t frames)
+{
+	uint32_t buff_frag = frames * source->channels - 1;
+	uint32_t curr_frames = frames;
+	int32_t prev_sum;
+	uint32_t channel;
+	int16_t *src;
+	int32_t sum;
+	uint32_t i;
+
+	for (i = 0; i < frames; i++) {
+		sum = 0;
+
+		for (channel = 0; channel < source->channels; channel++) {
+			src = audio_stream_read_frag_s16(source, buff_frag);
+			sum += *src;
+			buff_frag--;
+		}
+
+		/* first sign change */
+		if (i > 0 && (sum ^ prev_sum) < 0)
+			return curr_frames;
+
+		prev_sum = sum;
+		curr_frames--;
+	}
+
+	/* sign change not detected, process all samples */
+	return frames;
+}
+
+#endif /* CONFIG_FORMAT_S16LE */
+
+#if CONFIG_FORMAT_S24LE
+/**
+ * \brief Used to find nearest zero crossing frame for 24 in 32 bit format.
+ * \param[in,out] source Source buffer.
+ * \param[in] frames Number of frames.
+ */
+static uint32_t vol_zc_get_s24(const struct audio_stream *source,
+			       uint32_t frames)
+{
+	uint32_t buff_frag = frames * source->channels - 1;
+	uint32_t curr_frames = frames;
+	int64_t prev_sum;
+	uint32_t channel;
+	int32_t *src;
+	int64_t sum;
+	uint32_t i;
+
+	for (i = 0; i < frames; i++) {
+		sum = 0;
+
+		for (channel = 0; channel < source->channels; channel++) {
+			src = audio_stream_read_frag_s32(source, buff_frag);
+			sum += sign_extend_s24(*src);
+			buff_frag--;
+		}
+
+		/* first sign change */
+		if (i > 0 && (sum ^ prev_sum) < 0)
+			return curr_frames;
+
+		prev_sum = sum;
+		curr_frames--;
+	}
+
+	/* sign change not detected, process all samples */
```

Comment:
Probably best to wait til Monday for @singalsu comment too. 
This check should be per channel since each channel may cross zero at separate sample offsets.
e.g.
```c
/* return sample offset of ZC sample on channel */
int vol_get_zc_offset_near_fmt(stream, base_sample_offset, end_sample_offset, channel_stride)
```

Reply:
It would reduce the transitions more but it would make the copy function more complicated and increase MCPS so I proposed this simplified approach first. The remaining changes to handle per channel could be done later.